streaming/base/storage/upload.py

```diff
@@ -0,0 +1,330 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Write files to remote location which can be either Cloud Storage or a local path."""
+
+import logging
+import os
+import shutil
+import urllib.parse
+from tempfile import mkdtemp
+from typing import Any, Optional
+
+import tqdm
+
+__all__ = ['CloudWriter', 'S3Writer', 'GCSWriter', 'OCIWriter', 'LocalWriter']
+
+logger = logging.getLogger(__name__)
+
+
+class CloudWriter:
+    """Upload local files to a cloud storage."""
+
+    def __new__(cls,
+                local: Optional[str] = None,
+                remote: Optional[str] = None,
+                keep_local: bool = False,
+                progress_bar: bool = False) -> Any:
+        """Instantiate a cloud provider or a local writer based on remote location keyword.
+
+        Args:
+            local (str, optional): Optional local output dataset directory. If not
+                provided, a random temp directory will be used. If ``remote`` is provided,
+                this is where shards are cached before uploading. One or both of ``local``
+                and ``remote`` must be provided. Defaults to ``None``.
+            remote (str, optional): Optional remote output dataset directory. If not
+                provided, no uploading will be done. Defaults to ``None``.
+            keep_local (bool): If the dataset is uploaded, whether to keep the local dataset
+                shard file or remove it after uploading. Defaults to ``False``.
+            progress_bar (bool): Whether to display a progress bar while uploading to a cloud
+                provider. Defaults to ``False``.
+
+        Raises:
+            ValueError: Either local and/or remote path(s) must be provided
```

Comment:
trailing . to match elsewhere

Reply:
Addressed as part of https://github.com/mosaicml/streaming/pull/171/commits/5b5c10450637d4ccdbc8b5c2403a7d18f89b169b.